### ISSUE TITLE
How to log epoch's accuracy using Result object

### ISSUE THREAD

**Issue by sheffier** [PyTorchLightning/pytorch-lightning#3380]
#### What is your question?
Hi,

Before version #0.9.0, I used to log the predictions and targets at each step, and then at the "_epoch_end" method I aggregated the predication and targets and used these aggregations as input to the *Accuracy* metric to calculate the epoch's accuracy.

Is there a way to use the Result object in order to achieve the same? specifically, I would like to utilize the Result object convenience feature that allows me to only define a "_step" method without the necessity of defining "_epoch_end" method in the **LightningModule**

**Comment by ydcjeff** [comment]
Hi @sheffier , you can try with `log` method of `Result`
with `TrainResult`
```py
result = TrainResult(minimize=loss)
result.log('accuracy', acc, on_epoch=True)
```
with `EvalResult`
```py
result = EvalResult()
result.log('val_accuracy', val_acc, on_epoch=True)
```

Doc example:
https://pytorch-lightning.readthedocs.io/en/latest/results.html#logging
https://pytorch-lightning.readthedocs.io/en/latest/results.html#eval-logging

**Comment by sheffier** [comment]
Thanks @ydcjeff, but your suggestion assumes that I calculate the accuracy each step (because it will produce biased calculation when training on multiple GPUs).

My current solution is as follows:

```javascript
def tbptt_reduce_fx(inp):
    return inp[0]


class AccuracyLog:
    def __init__(self):
        self.acc = Accuracy(reduce_op="mean")

   def __call__(self, agg_in):
        pred_ids = torch.cat([item[0] for item in agg_in]).squeeze()
        target_ids = torch.cat([item[1] for item in agg_in]).squeeze()

        avg_acc = self.acc(pred_ids, target_ids)

        return avg_acc
```

```javascript
class MyLightningModel(LightningModule):
    def __init__(self):
        ...
        self.acc_log = AccuracyLog()

    def training_step(self, batch, batch_idx):
        outputs = self.forward(batch)
        targets = batch["target"]
         
        pred_ids = outputs.argmax(dim=1)
        
        ...

        acc_inputs = (pred_ids, targets)
        result.log(
            "train_acc",
            acc_inputs,
            on_step=False,
            on_epoch=True,
            prog_bar=True,
            reduce_fx=self.acc_log,
            tbptt_reduce_fx=tbptt_reduce_fx,
        )
```

Although this solution works, I don't like it. It feels like a hack. I would prefer a cleaner solution.

**Comment by s-rog** [comment]
This is quite unconventional and I don't think result objects are meant to be used this way. Your previous epoch_end implementation makes more sense to me.

**Comment by mlinxiang** [comment]
if i set result = pl.EvalResult() , then in val_epoch_end() can't use pl.EvalResult() again?
when i use the log in val_step() has no data write to tensorboard.

**Comment by agamemnonc** [comment]
> This is quite unconventional and I don't think result objects are meant to be used this way. Your previous epoch_end implementation makes more sense to me.
> 
> Edit:
> Though you can use results like this, since it's a dictionary
> 
> ```
> def training_step(self, batch, batch_idx):
>   ...
>   result.pred_ids = pred_ids
>   result.targets = targets
>   return result
> 
> def training_epoch_end(self, outputs):
>    pred_ids = outputs.pred_ids
>    targets = outputs.targets
>    # calculate acc
> ```

I have tried to use this approach, but the length of the saved values is equal to the batch_size instead of batch_size * num_batches, this indicating that only results from one step are saved rather than concatenated across steps.

**Comment by rhjohnstone** [comment]
I'm not sure my question is exactly the same, but it's related so I didn't want to open a new issue...

I'm computing a MSE loss, and using an `EvalResult` inside `validation_step` to log a validation loss after each training epoch. It has a default reduce function of `torch.mean`, but since I'm computing MSE loss of batches, the final batch will be smaller than the others. Therefore the overall average should be weighted to reflect this. Or is this weighting done somewhere that I just can't see?
